services/src/main/java/org/keycloak/protocol/saml/SamlService.java

```diff
@@ -517,6 +517,19 @@ public Response execute(String samlRequest, String samlResponse, String relaySta
             else
                 return handleSamlResponse(samlResponse, relayState);
         }
+
+        /**
+         * KEYCLOAK-12616, KEYCLOAK-12944: construct the expected destination URI in the same way that {@link KeycloakUriInfo#getAbsolutePath()}
+         * used to do prior to the changes introduced by the addition of the default hostname provider.
+         *
+         * @param session a reference to the {@link KeycloakSession}.
+         * @return the constructed {@link URI}.
+         */
+        protected URI getExpectedDestinationUri(final KeycloakSession session) {
+            KeycloakUriInfo info = session.getContext().getUri();
+            URI baseUri = info.getBaseUri();
+            return info.getAbsolutePathBuilder().scheme(baseUri.getScheme()).host(baseUri.getHost()).port(baseUri.getPort()).build();
```

Comment:
This is not going to work as Keycloak may be exposed on a different context-path on the reverse proxy. I.e. frontendUrl could be "https://mykeycloak.org/" while Keycloak is listening on "https://10.9.10.10:8443/auth".

Reply:
It would be better to just load the actual destination of the endpoints, rather than build it based on the request.

Reply:
See what @pedroigor did as an example: https://github.com/keycloak/keycloak/commit/7efaf9869acbb328f9dd8b159460cec7de6bc010#diff-19561d5afe444c1530b7f14bd01a7124R500

Reply:
@stianst first version of the fix was using the baseUri + destination of the endpoints. So if proxy is on https://mykeycloak.org/ the URL to be validated would be https://mykeycloak.org/realms/demo/protocol/saml. Reason I changed it is that the above is exactly what we used to so I thought users would be setting the destination with the /auth/ path anyway and changing it would "break" existing clients.

I will revert to my original impl in which the baseUri is composed with the endpoint destination.

Reply:
> This is not going to work as Keycloak may be exposed on a different context-path on the reverse proxy. I.e. frontendUrl could be "https://mykeycloak.org/" while Keycloak is listening on "https://10.9.10.10:8443/auth".

It does work as this is just the URL used for validating the destination. The redirection uses the baseUri so it properly redirects back to the proxy. I assumed users were setting the destination with the /auth/ path because of the way we used to buld the validation URL. I agree the destination should be set using solely the proxy path, so I will change that.

Reply:
Changed it so it now uses the destination of the SAML request endpoint. Test was improved as well, it now checks that the request fails if the frontendUrl isn't set to the proxy address.